packages/flutter/lib/src/gestures/resampler.dart

```diff
@@ -0,0 +1,424 @@
+      if (event.timeStamp > sampleTime) {
+        // Stop if event is not up or removed. Otherwise, continue to
+        // allow early processing of up and remove events as this improves
+        // resampling of these events, which is important for fling
+        // animations.
+        if (event is! PointerUpEvent && event is! PointerRemovedEvent) {
+          break;
+        }
+
+        // When this line is reached, the following two invariants hold:
+        // (1) `event.timeStamp > sampleTime`
+        // (2) `_next` has the smallest time stamp that's no less than
+        //     `sampleTime`
+        //
+        // Therefore, event must satisfy `event.timeStamp >= _next.timeStamp`.
+        //
+        // Those events with the minimum `event.timeStamp == _next.timeStamp`
+        // time stamp are processed early for smoother fling. For events with
+        // `event.timeStamp > _next.timeStamp`, the following lines break the
+        // while loop to stop the early processing.
+        //
+        // Specifically, when `sampleTime < _next.timeStamp`, there must be
+        // at least one event with `_next.timeStamp == event.timeStamp`
+        // and that event is `_next` itself, and it will be processed early.
+        //
+        // When `sampleTime == _next.timeStamp`, all events with
+        // `event.timeStamp > sampleTime` must also have
+        // `event.timeStamp > _next.timeStamp` so no events will be processed
+        // early.
+        //
+        // When the input frequency is no greater than the sampling
+        // frequency, this early processing should guarantee that `up` and
+        // `remove` events are always re-sampled.
+        final Duration nextTimeStamp = _next?.timeStamp ?? const Duration();
+        assert(event.timeStamp >= nextTimeStamp);
+        if (event.timeStamp > nextTimeStamp) {
+          break;
+        }
+      }
+
+      final bool wasTracked = _isTracked;
+      final bool wasDown = _isDown;
+
+      // Update pointer state.
+      _isTracked = event is! PointerRemovedEvent;
+      _isDown = event.down;
+
+      // Position at `sampleTime`.
+      final Offset position = _positionAt(sampleTime);
+
+      // Initialize position if we are starting to track this pointer.
+      if (_isTracked && !wasTracked) {
+        _position = position;
+      }
+
+      // Current pointer identifier.
+      final int pointerIdentifier = event.pointer;
+
+      // Initialize pointer identifier for `move` events.
+      // Identifier is expected to be the same while `down`.
+      assert(!wasDown || _pointerIdentifier == pointerIdentifier);
+      _pointerIdentifier = pointerIdentifier;
+
+      // Skip `move` and `hover` events as they are automatically
+      // generated when the position has changed.
+      if (event is! PointerMoveEvent && event is! PointerHoverEvent) {
+        dequeuedEvents.add(_clone(event, position, position - _position,
```

Comment:
If I understand the code correctly, only `PointerUpEvent` and `PointerRemovedEvent` is skipped, `PointerAddedEvent`, `PointerCancelEvent`, `PointerDownEvent`, `PointerEnterEvent`, etc, are all resampled with almost the algorithm with `PointerMoveEvent` and `PointerHoverEvent` as in `_samplePointerPositions`. I wonder why should we, and if we can have some part explain the resample algorithm in the doc of the class.

Reply:
And generally speaking, should we skip all events that we don't recognize/designed and only limit resampling to what we know?

Reply:
All events except `PointerAddedEvent` will be resampled given that we have future events queued that allow us calculate a resampled position. I've found that it's important to resample all events to ensure that the beginning and end of touch based animations are smooth. E.g. if we're not resampling `up` then it will appear as a sudden change in velocity when this event arrives unless the code is careful to take time stamps into account when determining velocity.

I explained that all events except 'added' are resampled in the doc of this class.